feat: Add classification metrics

josephmaa opened the pull request:
The aim of this PR is to provide more metrics besides accuracy that we can use for determining the effectiveness of a model. Currently adds auc roc curves.

josephmaa commented:
Example from test dataset:

<img width="1552" alt="Screen Shot 2022-04-05 at 4 33 16 PM" src="https://user-images.githubusercontent.com/37686936/161867780-4f97bbf7-cac1-4118-a169-5b61e5ca0952.png">

josephmaa commented:
Here are some of the results from printing the predicted outputs and the output of the decision function from the LinearSVC classifier:

```
[[0 0 0 0 0]
 [0 0 0 0 0]
 [1 0 0 0 0]
 [0 0 0 0 1]
 [0 0 0 0 1]
 [0 0 0 0 1]
 [1 0 0 0 0]
 [0 0 0 0 0]
 [0 0 0 0 0]
 [0 0 0 1 0]
 [0 0 0 1 0]
 [0 1 0 0 0]
 [0 1 0 0 0]
 [1 0 0 0 0]
 [0 0 0 0 1]
 [0 0 0 0 1]
 [0 1 0 0 0]]
[[-0.6488396  -0.96458057 -0.05070014 -0.33378502 -0.96120859]
 [-0.6488396  -0.96458057 -0.05070014 -0.33378502 -0.96120859]
 [ 0.91559418 -0.97292689 -0.97992985 -0.96112776 -0.97033759]
 [-0.9565376  -0.96457764 -0.97377236 -0.94916895  0.88494557]
 [-0.9565376  -0.96457764 -0.97377236 -0.94916895  0.88494557]
 [-0.9565376  -0.96457764 -0.97377236 -0.94916895  0.88494557]
 [ 0.91559418 -0.97292689 -0.97992985 -0.96112776 -0.97033759]
 [-0.6488396  -0.96458057 -0.05070014 -0.33378502 -0.96120859]
 [-0.6488396  -0.96458057 -0.05070014 -0.33378502 -0.96120859]
 [-0.95654125 -0.96457759 -0.97377657  0.8969803  -0.96121106]
 [-0.95654125 -0.96457759 -0.97377657  0.8969803  -0.96121106]
 [-0.9623366   0.63068604 -0.97727329 -0.6892753  -0.96637851]
 [-0.9623366   0.63068604 -0.97727329 -0.6892753  -0.96637851]
 [ 0.91559418 -0.97292689 -0.97992985 -0.96112776 -0.97033759]
 [-0.9565376  -0.96457764 -0.97377236 -0.94916895  0.88494557]
 [-0.9565376  -0.96457764 -0.97377236 -0.94916895  0.88494557]
 [-0.9623366   0.63068604 -0.97727329 -0.6892753  -0.96637851]]
```

josephmaa commented:
Unexpectedly, the fit using the prediction probabilities from the pyuoi `predict_proba` function yielded high values for the area under the roc-auc curves. (see below)

<img width="1552" alt="Screen Shot 2022-04-09 at 1 02 15 PM" src="https://user-images.githubusercontent.com/37686936/162589930-f7421b2d-e375-4ed3-86f5-a242ba0b3850.png">
<img width="1552" alt="Screen Shot 2022-04-09 at 1 02 36 PM" src="https://user-images.githubusercontent.com/37686936/162589974-a0a1115a-f4f7-4464-b2d7-1a3f3e3a7e05.png">
<img width="1552" alt="Screen Shot 2022-04-09 at 1 02 40 PM" src="https://user-images.githubusercontent.com/37686936/162589976-1f15fb97-2964-43ce-b0fc-66263c903f94.png">
<img width="1552" alt="Screen Shot 2022-04-09 at 1 02 43 PM" src="https://user-images.githubusercontent.com/37686936/162589977-fed2351e-939d-4102-a319-40fdccc57990.png">
<img width="1552" alt="Screen Shot 2022-04-09 at 1 02 46 PM" src="https://user-images.githubusercontent.com/37686936/162589981-002cd7e8-1252-4b53-b5a2-d70693ae75cb.png">

josephmaa commented:
Have generated the precision-recall curves with the LinearSVC model. I also increased the number of samples since the first dataset didn't have enough examples of the rarer events to generate useful precision-recall curves. 

<img width="1552" alt="Screen Shot 2022-04-09 at 1 30 16 PM" src="https://user-images.githubusercontent.com/37686936/162590777-c3d544df-ac07-489e-b96f-dde6bda8efaa.png">
<img width="1552" alt="Screen Shot 2022-04-09 at 1 30 28 PM" src="https://user-images.githubusercontent.com/37686936/162590778-8ad76815-7905-47d0-99b3-7b0bc703cd59.png">
<img width="1552" alt="Screen Shot 2022-04-09 at 1 30 32 PM" src="https://user-images.githubusercontent.com/37686936/162590780-aec4d397-d1d4-425f-9243-3db1345b07bf.png">

shwan19 commented:
A few questions:
1) What was in the input dataset you used for the linear SVC? (I'm guessing not the behavior data?)
2) How many positive examples are there of class 1 v. class 2 v. class 3 in the PR plots?
3) Why are there 4 classes for ROC plots and 3 classes for PR plots?

But I think there is a straightforward answer to your question about the class 3 PR plot. 1 edge v. more than 1 edge is not the central point, I don't think. Take a look at the y-axes for class 1 v. class 2 v. class 3. You'll notice that the lowest value on the y-axis is very quite different across these three plots. The remaining portion of the plot (between y = 0 and y = min) is implied/hidden (though of course you can implement plotting differently to get the whole plot). The AUPRC is thus the area between each PRC and y = 0 for x = [0,1].

josephmaa commented:
> A few questions:
> 
> 1. What was in the input dataset you used for the linear SVC? (I'm guessing not the behavior data?)

The input dataset was generated with the `--use_small_dataset` flag, so not the behavior data

```
if parsed_args.use_small_dataset:
        n_features = 20
        n_inf = 10
        x, y, _, _ = make_classification(n_samples=1000,
                                         random_state=10,
                                         n_classes=5,
                                         n_informative=n_inf,
                                         n_features=n_features,
                                         shared_support=True,
                                         w_scale=4.)

        x_train, x_test, y_train, y_test = train_test_split(
            x, y, random_state=parsed_args.training_seed)
```

> 2. How many positive examples are there of class 1 v. class 2 v. class 3 in the PR plots?

Here's the confusion matrix of values:

<img width="1552" alt="Screen Shot 2022-04-10 at 2 07 22 AM" src="https://user-images.githubusercontent.com/37686936/162611118-68b47f65-006d-4e18-a542-2a80793ff3a4.png">

I added some code so that we can visualize the results for all subsequent runs.

> 3. Why are there 4 classes for ROC plots and 3 classes for PR plots?

That's a good catch. There was a bug in the code that I think was related to the ordering of the plots getting generated. The final plot is here. You can see in the commit after this comment that I changed the code to plot the precision-recall curves first.

> But I think there is a straightforward answer to your question about the class 3 PR plot. 1 edge v. more than 1 edge is not the central point, I don't think. Take a look at the y-axes for class 1 v. class 2 v. class 3. You'll notice that the lowest value on the y-axis is very quite different across these three plots. The remaining portion of the plot (between y = 0 and y = min) is implied/hidden (though of course you can implement plotting differently to get the whole plot). The AUPRC is thus the area between each PRC and y = 0 for x = [0,1].

That is a good point. I think some articles mention plotting the no-skill line "The no-skill line changes based on the distribution of the positive to negative classes. It is a horizontal line with the value of the ratio of positive cases in the dataset.". If we assume the no-skill line to pick a random class, since the dataset is fairly evenly distributed, I would expect the line to be plotted at approx. 0.2 on the y-axis for these precision-recall curves. I'm still thinking about how we would have a no-skill line based on the classifier picking only one class (the alternative[ mentioned in the article](https://machinelearningmastery.com/roc-curves-and-precision-recall-curves-for-classification-in-python/#:~:text=A%20precision%2Drecall%20curve%20is,constant%20class%20in%20all%20cases.)).

kdquine commented on classifier/roc_curve.py:
```diff
@@ -0,0 +1,52 @@
+    df.sort_values('confidence', ascending=False)
+
+    # Accumulate values and calculate tpr and fpr rates per element
+    df['tpr'] = 0
```
this looks unfinished

josephmaa replied:
removed

josephmaa commented:
Did some testing with the `auc` function to test its functionality on multi-class classification since that was one of the questions that we had with it. Using the values from [the documentation](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.auc.html), I tested alternative use of binary labeling prior to getting passed into `fpr` and `tpr` which worked well.

![Screen Shot 2022-07-12 at 2 40 44 PM](https://user-images.githubusercontent.com/37686936/178600718-7e4b8f97-97a6-48fe-a2ee-9b51b7916809.png)

josephmaa commented:
Here are some real-data from a single class within a PC:
<img width="752" alt="Screen Shot 2022-07-12 at 4 44 49 PM" src="https://user-images.githubusercontent.com/37686936/178618227-488ae212-10a7-443d-a7ff-b88fb3860d08.png">
<img width="752" alt="Screen Shot 2022-07-12 at 4 44 43 PM" src="https://user-images.githubusercontent.com/37686936/178618232-b9e1d941-57fd-485f-aaff-92ac41dbf3d4.png">
<img width="752" alt="Screen Shot 2022-07-12 at 4 44 37 PM" src="https://user-images.githubusercontent.com/37686936/178618235-b87b17b6-8c20-4b5b-b7b8-4095d7ad9c5a.png">
<img width="752" alt="Screen Shot 2022-07-12 at 4 44 33 PM" src="https://user-images.githubusercontent.com/37686936/178618236-c9276fc2-644d-470a-aca0-185ad3fd432d.png">
<img width="752" alt="Screen Shot 2022-07-12 at 4 44 28 PM" src="https://user-images.githubusercontent.com/37686936/178618239-75760c8a-ea7a-4b7d-9953-e9c4d5752570.png">
